plugin/pkg/admission/nodeenv/admission.go

```diff
@@ -0,0 +1,234 @@
+	var namespace *api.Namespace
+
+	namespaceObj, exists, err := p.store.Get(&api.Namespace{
+		ObjectMeta: api.ObjectMeta{
+			Name:      nsName,
+			Namespace: "",
+		},
+	})
+	if err != nil {
+		return errors.NewInternalError(err)
+	}
+	if exists {
+		namespace = namespaceObj.(*api.Namespace)
+	} else {
+		namespace, err = p.defaultGetNamespace(nsName)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return err
+			}
+			return errors.NewInternalError(err)
+		}
+	}
+	nsTolerations, err := p.getNameSpaceTolerations(namespace)
+	if err != nil {
+		return err
+	}
+
+	podTolerations, err := api.GetTolerationsFromAnnotations(pod.Annotations)
+	if err != nil {
+		return err
+	}
+
+	if conflictingTolerations(nsTolerations, podTolerations) {
+		return fmt.Errorf("namespace tolerations and pod tolerations conflict")
+	}
+
+	// modified pod tolerations = namespace tolerations + current pod tolerations
+	podTolerations = mergeTolerations(nsTolerations, podTolerations)
+
+	tolerationStr, err := json.Marshal(podTolerations)
+	if err != nil {
+		return err
+	}
+
+	pod.Annotations[api.TolerationsAnnotationKey] = string(tolerationStr)
+	return nil
+
+}
+
+func NewPodNodeEnvironment(client clientset.Interface, clusterDefaultTolerations []api.Toleration) *podNodeEnvironment {
+	// TODO: make it a shared cache to use between admission plugins
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return client.Core().Namespaces().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Core().Namespaces().Watch(options)
+			},
+		},
+		&api.Namespace{},
+		store,
+		0,
+	)
+	return &podNodeEnvironment{
+		Handler: admission.NewHandler(admission.Create),
+		client:  client,
+		store:   store,
+		clusterDefaultTolerations: clusterDefaultTolerations,
+		reflector:                 reflector,
+	}
+}
+
+func (p *podNodeEnvironment) defaultGetNamespace(name string) (*api.Namespace, error) {
+	namespace, err := p.client.Core().Namespaces().Get(name)
+	if err != nil {
+		return nil, fmt.Errorf("namespace %s does not exist", name)
+	}
+	return namespace, nil
+}
+
+func (p *podNodeEnvironment) getNameSpaceTolerations(namespace *api.Namespace) ([]api.Toleration, error) {
+	nsTolerations, err := api.GetTolerationsFromAnnotations(namespace.Annotations)
+	if err != nil {
+		return nil, err
+	}
+
+	if nsTolerations == nil {
+		nsTolerations = p.clusterDefaultTolerations
+	}
+
+	return nsTolerations, nil
+}
+
+// conflictingTolerations returns true if the key of two tolerations match
+// but one or more other fields differ, otherwise returns false
+func conflictingTolerations(first []api.Toleration, second []api.Toleration) bool {
+	for i := range first {
+		for j := range second {
+			if first[i].Key == second[j].Key {
+				if first[i].Operator != second[j].Operator ||
```

Comment:
Isn't there some (partial) order of tolerations and merging should take the maximal toleration? E.g.:

- namespace: `<"env", "prod", "">`, equal
- pod: `<"env", "", "">`, exists

This means that all pods in the namespace tolerate "env=prod" taints. But the given pod even tolerates any "env" taints, e.g. also "env=test".